### ISSUE TITLE
[RFC] Move to Typescript

### ISSUE THREAD

**Issue by HenriBeck** [cssinjs/jss#1035]
With more OS projects now planning to move to Typescript (jest, Vue.js, etc.), I think it would be time to think about doing the same. Typescript has really matured and now with eslint and babel both supporting Typescript as well, it should be future proof.

Additionally, I think the tooling around Typescript (speed, autocompletion) etc. is far head of flow currently.

I would start with "rewriting" smaller packages (like `css-functions` and `theming`) in the beginning.
Also, I started writing the new docs with Typescript.

__Are you willing to implement it?__
Yes

**Comment by TrySound** [comment]
Besides better tooling I saw a lot of unsoundness in TS which is "won't fix" when flow to gonna be and actually already safer than TS in some cases. For me it doesn't worth to switch.
I'm quite happy with flow in big project and open source.

**Comment by HenriBeck** [comment]
Why would Typescript be less type-safe?

**Comment by TrySound** [comment]
Here guys from flow team provides some examples
https://discord.gg/zGY9gW

For example TypeScript doesn't have exact types at all (https://github.com/Microsoft/TypeScript/issues/12936) when flow is gonna have them by default. Object spread is unsound (https://github.com/Microsoft/TypeScript/issues/15454) when flow is already more sound here and is gonna improve experience with exact by default.

![image](https://user-images.githubusercontent.com/5635476/52887392-949ad880-3188-11e9-89af-f1b10188dc2a.png)

**Comment by TrySound** [comment]
Flow has some problems but flow team is quite active and open last few weeks. They are going to the right direction but their plan is time consuming. Now they have resources to work with community.

**Comment by kof** [comment]
That's interesting, do you know how many people work on flow full time? I feel like flow with strict mode and better error descriptions would actually make the whole thing actually quite competitive. It's just a matter if companies are willing to invest enough to improve.

**Comment by TrySound** [comment]
This is just now. There are more people there
![image](https://user-images.githubusercontent.com/5635476/52889438-fe1ee500-3190-11e9-9fac-0aa03025e462.png)

**Comment by kof** [comment]
Fulltime contributors?

**Comment by TrySound** [comment]
It's a chat with flow team (facebook guys). The link I put above.

**Comment by HenriBeck** [comment]
I think flow error descriptions and reporting is quite better (especially when it comes to react props) than Typescripts.

What I now dislike the most about it is how slow the editor integration is and that types are an afterthought.

**Comment by TrySound** [comment]
Which editor do you use?

**Comment by HenriBeck** [comment]
Webstorm, and for me the autocompletion, auto-import etc. is really slow, I think in a recent release they improved the language server to handle request while they are recomputing all of the types though.

**Comment by TrySound** [comment]
There is one representative from webstrom team in flow chat. They are discussing better integration. Yeah, main problem is checking too much. Ignoring things usually helps.

**Comment by HenriBeck** [comment]
The reason why I proposed this change that a lot of quite large projects are moving into this direction.
And yeah, TS maybe is sometimes unsafer, but I think this only applies for certain type definitions/operators.

The reason I got to like Typescript so much is that the language was designed with types in mind and not just an afterthought/addition like flow.

**Comment by TrySound** [comment]
Well, sorry, but herd instinct is a bad reason to refactor all the project. It looks like "flow is a shit and nobody will ever use, let's kill it". By doing this you just add fuel to the fire of hate which is really started by just one person.

By the way about big projects decisions. Here's another one
https://github.com/mdn/mdn/pull/53/files#diff-992ca475abeb0c101165f19ff62d37e7R14

**Comment by TrySound** [comment]
And typescript was designed around the same javascript with all it's problem. It's not really different language. Reason/ocaml or haskell was designed with types.

**Comment by kof** [comment]
They seem to want to use flow without types or adding types gradually.


I think the goal for a library should be the highest type safety possible.

**Comment by HenriBeck** [comment]
I never said that Typescript is a different language and has not the problems of javascript, but at the same time Typescript was designed around types (which in my opinion solves a few of them). The best comparison for that in my mind is how you import/export types in Typescript vs. flow.

**Comment by TrySound** [comment]
Flow was also designed around types. It's type system.

Well, I dont see any sense in import way. It's just a choice without "not types in mind"

Flow coming types first arch name is quite descriptive.

**Comment by HenriBeck** [comment]
What I mean by "designed around types" is that flow tries to add types on top of javascript, and Typescript is a superset of javascript which compiles down to js.

**Comment by kof** [comment]
That just means that flow is not going to add language features, just types, where typescript has done so and might do more. In that sense, ts has more power. At the same time, the more ts does so, the further away it is from js.

Most things other than that are basically implementation details and features which one has but another one doesn't and vise versa.

I think we need to clearly define criteria for our choice first of all.

**Comment by TrySound** [comment]
Its not true. Both add types and invalid syntax to javascript. Both are compiled to javascript. But flow team just is a bit more careful with JS semantics.

**Comment by kof** [comment]
Where did flow add something that compiles to runtime javascript?

**Comment by TrySound** [comment]
Well exact types is a good reason to stay with flow.

**Comment by TrySound** [comment]
What babel compiles to JS from TS?

**Comment by kof** [comment]
For example
```js
enum Direction {
    Up = 1,
    Down,
    Left,
    Right,
}
```

**Comment by kof** [comment]
> Well exact types is a good reason to stay with flow.

Can you expand pls? Afaik both have this, but mb you know more than I do

**Comment by kof** [comment]
TBH my biggest problem with both flow and TS is to maintain types for both. Ideally I want to do it only once.

**Comment by TrySound** [comment]
Read the thread. TS doesnt have exact types.

**Comment by TrySound** [comment]
AFAIK enums are not compiled via babel.

**Comment by TrySound** [comment]
@kof Flow team is working on definitions generator which will allows simply map them to TS typings.

**Comment by HenriBeck** [comment]
But TS already has it, and yeah maybe TS is sometimes a little bit less type-safe but so is flow in some cases.

So, in the end, I think it comes down to these points:
- [ ] Which gives up the best experience as the maintainers?
- [ ] Are people willing to maintain it (are we still going to have external contributors)
- [ ] Which choice is the more future proof

Feel free to add points to the list; I might have missed some.

**Comment by kof** [comment]
- [ ] Maintaining 2 type systems is always pain in the ass
- [ ] So far most useful contributions have been issues and rarely tests. So it doesn't even matter what language we use internally.

**Comment by jeremy-coleman** [comment]
https://github.com/yarnpkg/berry/tree/master/packages 
Yarn v2,
Rxjs, vue, angular, inferno, Mobx , all ts. Typescript is a better choice. Although really , you should look at your dependants and make the choice based on that. Fyi after retyping core ive found several problems on the current types, as with most flow libraries. If you do decide to switch to typescript, i will gladly help, it wont take long

**Comment by TrySound** [comment]
@jeremy-coleman Retyping to flow may also catch errors. Both type systems have own features.

**Comment by kof** [comment]
We have ts types for public api, feel free to submit your fixes if they can
improve those for now.

**Comment by jeremy-coleman** [comment]
I mostly changed source code in core only and integrated with glamor jss. , i dont think its usable outside my use case. Honestly though, csstype and the built in dom types have already done most the work, for me its easier to start from a js base than flow and focus on adding types to the input params (which is super easy if you already understand how the code works ) . There are also some codemods for flow to ts , havnt tried them out though. Jss is in a tough spot though bc of the plugin system. It would probably be better to create a second repo for the ts version like jss-ts or something

**Comment by HenriBeck** [comment]
Creating a separate version just for TS isn't an option as we would need to do bug fixes etc. twice.

**Comment by kof** [comment]
@jeremy-coleman you mean create a repo with a specific plugins setup that imports from main repo and  exposes types or did you suggest to copy the entire codebase?

**Comment by worudso** [comment]
@kof @jeremy-coleman 
https://github.com/cssinjs/jss/issues/1036
As I mentioned in the last comment in that issue, I suggest making `createStyleSheet` generic to support typings for plugins, providing typings for vanilla jss as default.
```typescript
// it only allows vanilla jss's style definition
jss.createStyleSheet({
  // ...
})

import { NestedStyle } from "jss-plugin-nested"; // it's not there now.
import { ObservableStyle } from "jss-plugin-rule-value-observable";
jss.createStyleSheet<NestedStyle & ObservableStyle>({
  // ...
})
```
The current type definition under devleopment partially supports plugins listed on https://cssinjs.org/.
It supports `jss-plugin-rule-value-observable` but doesn't support `jss-plugin-nested.`
And what about 3rd-party plugins? Is it ok that they are left untyped and force programmers to type `any`?

**Comment by jeremy-coleman** [comment]
https://github.com/jeremy-coleman/jss

**Comment by HenriBeck** [comment]
I think both type systems have their features and flaws. Both of them are not perfect, but I believe Typescript would give us a better DX than flow.

Also, I just noticed that in flow you couldn't extend interfaces which I find a little bit weird.

**Comment by TrySound** [comment]
I feel object type spread with exact types is more natural than extending interfaces.

**Comment by oliviertassinari** [comment]
> You cannot prove any interesting properties about TypeScript code without soundness, so why did the designers intentionally break soundness?

> TypeScript challenges the status quo, claiming that lowering the cognitive overhead of using types is more important than type soundness. They argue that JavaScript programmers, their main audience, would find it easier to use TypeScript if they deviated from traditional norms.

Pragmatism? *https://frenchy64.github.io/2018/04/07/unsoundness-in-untyped-types.html*

**Comment by HenriBeck** [comment]
I just tried upgrading to a newer flow version, and we had problems with our `memoize` function, so I tried to fix it that it's still type-safe and at the same time generic. I couldn't find a solution, because there is no way of defining overload types for functions.

**Comment by HenriBeck** [comment]
Also, I wasn't able to get `[declarations]` in the config working to ignore a new error inside the `node_modules`

**Comment by HenriBeck** [comment]
I would prefer it if we did an incremental switch to typescript. I think the developer experience is far inferior to Flow and we might even see more engagement from the community in terms of PRs.

**Comment by rob2d** [comment]
$.02: you can use Typescript in ESDoc with VSCode and get all of the annotations/auto predictions and linting without wasting time you could be using on more productive things... Like developing and making sure there's good integration tests (followed by unit tests if life is luxurious, but good developers think in unit tests while they code as habit).

**Comment by jeremy-coleman** [comment]
@HenriBeck i think you meant the dev experience is far superior? :P

**Comment by Saturate** [comment]
> I would prefer it if we did an incremental switch to typescript. I think the developer experience is far inferior to Flow and we might even see more engagement from the community in terms of PRs.

I would like to contribute to this change. Is there any plan going forward?

I'm thinking if we can add the babel typescript plugin, it should be possible to do it incremental and rewrite modules pretty easily.

**Comment by kof** [comment]
Another option I discovered today, also used by preact https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html

Checking types using jsdoc types using ts compiler.

**Comment by karlhorky** [comment]
> Checking types using jsdoc types using ts compiler.

Yeah, this is a good way to do an incremental migration by building up the types and soundness in your existing JS files.

**Comment by hosseinmd** [comment]
By Jsdoc we can't define some complex types, But we could use Jsdoc to documentation.

**Comment by hosseinmd** [comment]
Migration to typescript could save us from some bugs in compile time or when we're coding.

**Comment by kof** [comment]
Complex types can be defined in type files and used inside comments, right?

**Comment by hosseinmd** [comment]
Yes but what is the benefit of that, why we don't use typescript?

**Comment by hosseinmd** [comment]
I found some type which is different in .d.ts vs source, like `InsertionPoint`.

**Comment by hosseinmd** [comment]
RuleListOptions is very different

**Comment by kof** [comment]
> Yes but what is the benefit of that, why we don't use typescript?

The benefit is avoiding having types mixed with js code, it's easier to read for people who are not using ts in every day work. 
A side effect of that is also js code can still be parsed directly by any parser or even run directly in the browser as is if needed. It's just javascript in this case.  I am sure webpack maintainers have more to say about this approach since they use it for years.

**Comment by hosseinmd** [comment]
This is not a clear, through this way developers should understand jsdoc and typescript both.

**Comment by karlhorky** [comment]
After starting an increasingly large project with JSDoc, I would recommend staying away from JSDoc. It's a different, less ergonomic syntax (then you need to learn both if you want to have the complex types in TS files as well), you can't do everything you can in TypeScript, other downsides too - TypeScript is so much simpler and more natural. The downsides of a learning curve for TS beginners are far outweighed by the downsides of JSDoc, in my opinion.

**Comment by hosseinmd** [comment]
@kof should the fallbacks add to types of JssStyles? Is it missed, or is a reason?

**Comment by kof** [comment]
@hosseinmd lets not dicuss specific type problems here, it's out of scope, gonna remove this message as well. Fallbacks clearly belong to the type def.

**Comment by hosseinmd** [comment]
Tests can move to ts too. We should discuss it too.

**Comment by ITenthusiasm** [comment]
Figured I'd finally add some thoughts. I might have more later; my time is limited.

I think it's good to recall attention to the question of "What most enhances the developer experience?" It's probably the most important one, though it may fight with the question of whether or not people are willing to maintain the code in first place.

To that extent, I do think that migrating to TS in the long term is a significantly better choice.

### Popularity
I could be wrong, but I feel like on average there are going to be less developers familiar with flow. As HenriBeck (and jeremy-coleman) pointed out, many _popular_ libraries have moved their codebases to TS.

> Well, sorry, but herd instinct is a bad reason to refactor all the project.

But herd instinct isn't the only matter here. If popular libraries are not only used by several people but _contributed to_ by several people, this theoretically implies that more contributors (especially the good ones) would be more familiar with TS than with flow (_theoretically_). If we want helpful contributors over the long haul, this is worth considering.

I don't consider this the strongest point.

### Ease of Access
Flow honestly kind of feels like a hassle. The flow server is always crashing on my IDE. And even getting flow setup in something like VS Code is... a weird experience. I raised #1461 to try to help that problem, but the reality is that contributors shouldn't even have to deal with this in the first place.

TypeScript works straight out of the box, and it's fast. This is somewhat worth considering.

### Maintainability and Tooling

>  Maintaining 2 type systems is always pain in the ass

If we use TypeScript, we wouldn't have to worry about this problem. We could write the codebase in TypeScript, and when we do the build, the type definitions could be automatically generated. This is **huge** because we don't have to worry about keeping JS (or JS + Flow) in sync with the TS types we expose. So the codebase becomes smaller, and the potential to run into problems like #1482 gets eliminated.

As far as I understand (correct me if I'm wrong), for users to get type definitions for a library, they _need_ TypeScript [definition] files. This means we'll have to be working with TypeScript anyway to some extent or another. We can avoid duplicating our work by leveraging TS.

### Miscellaneous

> The benefit is avoiding having types mixed with js code, it's easier to read for people who are not using ts in every day work.

There's actually a kind of double-edged sword when it comes to this. If you have a small, simple repository that's very easy to contribute to, then there's a chance you don't need TypeScript. Someone can just read the code without any distractions, and life is simple. Some parts of the Testing Library family are an example of this.

However, as the codebase expands, having to deal with raw JS actually gets worse. You end up with a lot of functions and a lot of moving pieces that are hard to follow. Types help reduce some of the cognitive load here, and it seems to some extent we all agree that they're necessary. The benefit of using Flow or JSDocs is that people can technically get away with using raw JS and ignoring types.

However, the reality is that contributors probably shouldn't be adding code to a large project without [good] type information because it will cause more ambiguity/confusion for people further down the road. So the contributor will have to learn JSDocs/Flow/TypeScript anyway if we're looking at a code base that's more maintainable long term.

**Comment by ITenthusiasm** [comment]
I really do think TS is the better play in the long haul.

I'm not sure about what the migration path should look like. And I don't know how quickly the work should be done.

Something gradual and non-disruptive would probably be best. (But that's assuming everyone decides migrating to TS is a good idea.)

**Comment by ITenthusiasm** [comment]
> I started rewriting the core in my fork in Typescript as an experiment.

@HenriBeck I know it's been a while, but what's the status of that fork?

**Comment by kof** [comment]
> Ah, I see. I may have phrased things poorly. When I referred to types getting generated automatically, I was referring to TS type definitions (because that's what the end user will need to get intellisense if I understand correctly). So if we write the entire codebase in TS, we won't need separate type definitions files. Instead, during the build process, the TS type defs that the end users need would automatically be generated.

Yeah, this doesn't remove the need for all the flowtype users to have their types.

**Comment by kof** [comment]
> If JS/TS absolutely had to be separated, then yeah we could add tests for each of the types in addition to the implementation details.

Wait that's exactly the point I am making, we don't have to. We have real tests. If those tests are writen in TS, they already use the interface and validate it.

**Comment by kof** [comment]
> But separation increases cognitive load

Isn't intellisense showing all the types when you hover over functions/classes? I mean this way we still see the types and we don't need to see them if we don't want to! That's ideal from both worlds. I am not sure if it's possible to link a typedef to the module specifically without making a reference? But I guess we could make the references in the code files

**Comment by kof** [comment]
I tried to do that, so far this has worked:

https://codesandbox.io/s/admiring-goldwasser-27rki?file=/main.js

I have seen people saying on SO it's possible to automatically match type defs and code, do you know how? Didn't work for me.

**Comment by karlhorky** [comment]
If supporting Flow users is an absolute necessity, maybe it's worth taking a look at `flowgen`?

https://github.com/joarwilk/flowgen

**Comment by kof** [comment]
@karlhorky we can try, but I wouldn't bet on it to work properly considered complexity of TS types we currently have, including imported CSS types.

**Comment by kof** [comment]
Autocomplete the way I used it here doesn't show a full type def for the function when asking it to show the type. Jumping to type definition works. This approach wouldn't be nice for a product team, but considered this is a lib, it's doable. I would love though to not have to add `@type` and to see a full type def on hover. To me this would be perfect.

https://codesandbox.io/s/admiring-goldwasser-27rki?file=/main.js


https://user-images.githubusercontent.com/52824/113268821-1baf5a00-92d8-11eb-9e2e-786087775087.mp4

**Comment by hosseinmd** [comment]
jsdoc cannot Solve problem of flow. if we use ts/jsdoc user can't test real flow. 

But about your example
I have used JsDoc before, I believe that typescript is more clear than Jsdoc. For example if your function have generic params, you should define them in jsdoc too.

Compare your code in js and ts version

Your example
```js
// @ts-check
/** @type {import("./main").Fn}*/
export const fn = () => 1;
```
ts version of that is:
```ts
export const fn = (): string => 1;
```

**Comment by hosseinmd** [comment]
If we want to use jsdoc, we should code more without a good reason or benefit.

**Comment by kof** [comment]
@hosseinmd do you know the answer to the question I posted?

**Comment by hosseinmd** [comment]
@kof excuse me what's the question?

**Comment by hosseinmd** [comment]
> I have seen people saying on SO it's possible to automatically match type defs and code (so that we don't have to do @type import ... next to every function), do you know how? Didn't work for me.

I don't know

**Comment by kof** [comment]
If we have to deal with more writing and limited syntax I wouldn't want to use it. I am trying to understand what happens if JS files are linked to type defs. Type defs have a full TS support if I understand correctly. If they can be linked in a nice way to code and vscode can show the types on hover - I don't see any problems, but there are 2 "ifs" here.

**Comment by ITenthusiasm** [comment]
> Yeah, this doesn't remove the need for all the flowtype users to have their types. Users using flowtype in their code base won't be able to use TS type defs to typecheck. VSCode can give them the intellisense, true.

Ah, okay that's super helpful! That was the disconnect for me. Yeah...that makes this a little hard... hm...

> Wait that's exactly the point I am making, we don't have to. We have real tests. If those tests are writen in TS, they already use the interface and validate it.

Yes! Originally, it wasn't really my preference to have that separated out. However, now that I know that flow users still need flow type definitions, I'm not sure what's best right now.


I need to think more about the problem now. Do we compile flow type definitions in a separate file or do they remain with the code?

**Comment by kof** [comment]
both options exist, we currently don't generate type defs, we generate a .flow file which is `export * from '../src';`, but we can as well generate type defs. I think it's the same thing as with TS

**Comment by HenriBeck** [comment]
Hey Guys,

I haven't been active in the project for a long time but I hope to get back into it and find some time in the coming weeks.

My take is this by now on the whole Flow vs. Typescript topic:

At this point, it is clear and a fact that TS has the upper hand, and Flow will not catch up at this point (just looking at the installs on NPM and usage in the industry).
Considering this, I don't see an argument against migrating the codebase to 100% TS, it is more used in the community and the better-supported ecosystem/community.

I guess the only question is if we want to start out with providing Flow types or see if this is needed at all.

Overall the challenge will still be to provide a good CSS style definition due to the very dynamic nature of our plugin system.

**Comment by kof** [comment]
I think moving public interfaces which are currently part of code to type defs file isn't too hard, much harder is to make flow types same way detailed, with CSS types and stuff. We can just wait if someone is interested to contribute and otherwise not waste the energy on improving them. Just keep them as they are and move them out of the js files.

**Comment by kof** [comment]
I made the first step and moved flow types to a typedef file for one module https://github.com/cssinjs/jss/pull/1509
It works incrementally, I could merge just this one. We could merge them all at once though, I think it's not too hard.

Anybody wants to take on some modules and do the same thing?

**Comment by hosseinmd** [comment]
I would like to move all react-jss to ts. can I do it?

**Comment by kof** [comment]
> I like to move all react-jss to ts. can I do it?
yes, though first we need to move flow definitions to .flow files, like I showed above :) 

feel free to send a PR right onto that branch move-flow-to-typedefs

**Comment by kof** [comment]
I finally finished moving all flow types in all packages, once this is merged, nothing stops us from using ts https://github.com/cssinjs/jss/pull/1509

**Comment by kof** [comment]
This is merge https://github.com/cssinjs/jss/pull/1509

**Comment by hosseinmd** [comment]
Could you mention a Road map for migrate to typescript?

ّWe need to pursue this more seriously
